incremental builds: add derivation override functions

messemar opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

messemar committed:
Checkpointedbuilds: add derivation override functions

One can use this functions to allow incremental builds of derivations

Lassulus commented on pkgs/build-support/build-incremental.nix:
```diff
@@ -0,0 +1,37 @@
+rec {
+  /* Prepare a derivation for local builds.
+    *
+    * This function adds an additional outout for a derivation,
```
```suggestion
    * This function adds an additional output for a derivation,
```

messemar replied:
fixed

messemar committed:
checkpointedBuilds: add usage example based on virtualbox

roberth commented:
Intriguing!

Some thoughts:
 - buildOut: "out" is a little redundant
 - the intermediate files are more likely not to be reproducible than normal outputs
 - most of all: Please write tests

messemar committed:
checkpointedBuilds: rename buildOut to checkpointedBuildArtifacts

messemar committed:
checkpointedBuild: add checkpointed build test based on pkgs hello

messemar commented:
> Intriguing!
> 
> Some thoughts:
> 
>     * buildOut: "out" is a little redundant
> 
>     * the intermediate files are more likely not to be reproducible than normal outputs
> 
>     * most of all: Please write tests

Thanks for your input.

I have changed the names and added a test. Let me know If there is more to do for testing.

To your second concern: 
I guess the intermediate files should be as reproducible as normal outputs, as they depends on the input and the compiler, also.
But I can not guarantee that they will be reproducible at any time.
I run the same derivation several times, and the derivation would not be rebuilt if nothing in the source tree changes.

tfc commented:
This is pretty cool.

Interface wise i think it might be nicer to use and it seems like nothing speaks against doing it like this maybe?

```nix
incrementalDrv = pkgs.buildIncremental drv;
```

this could be done by implementing `build-incremental.nix` like this:

```nix
{ pkgs }:
let
  prepareIncrementalBuild = drv: ...;
  mkIncrementalBuild = drv: ...;

in

drv: previousArtifacts:

mkIncrementalBuild (prepareIncrementalBuild drv) previousArtifacts
```

The tests would look slightly different but would show the same thing.

messemar commented:
> This is pretty cool.
> 
> Interface wise i think it might be nicer to use and it seems like nothing speaks against doing it like this maybe?
> 
> ```nix
> incrementalDrv = pkgs.buildIncremental drv;
> ```
> 
> this could be done by implementing `build-incremental.nix` like this:
> 
> ```nix
> { pkgs }:
> let
>   prepareIncrementalBuild = drv: ...;
>   mkIncrementalBuild = drv: ...;
> 
> in
> 
> drv: previousArtifacts:
> 
> mkIncrementalBuild (prepareIncrementalBuild drv) previousArtifacts
> ```
> 
> The tests would look slightly different but would show the same thing.

You can not handle it that way, because there is no possibility to detect an earlier build automaticly, as nix is lacking the corresponding Support.

tfc commented:
> You can not handle it that way, because there is no possibility to detect an earlier build automaticly, as nix is lacking the corresponding Support.

Ok sorry for the noise. Now i get how it works.

roberth commented:
- What happens when a source file is removed in the new build? It seems unavoidable that the accompanying intermediate file will be present, which is a risk, but a decent build script will only glob for source files.
This needs a test case.

 - In the existing test case, could you check that unmodified files aren't recompiled?

messemar commented:
> * What happens when a source file is removed in the new build? It seems unavoidable that the accompanying intermediate file will be present, which is a risk, but a decent build script will only glob for source files.
>       This needs a test case.
 Yes, good Idea, I will add it
>     * In the existing test case, could you check that unmodified files aren't recompiled?

Yes.

messemar commented:
I've added the requested tests.

roberth commented on pkgs/test/incrementalBuild/default.nix:
```diff
@@ -0,0 +1,57 @@
+    '';
+  });
+
+  incrementalBuiltHelloWithRemovedFile = buildIncremental.mkIncrementalBuild patchedHelloRemoveFile baseHelloRemoveFile.incrementalBuildArtifacts;
```
Please check that the additional file is not present during the incremental `buildPhase`.

`hello` doesn't really suffer from this potential issue, because it doesn't seem to glob for source files, so you could either test another derivation that does scan its source directory (example: `cp *.md $out`) or you could override the `buildPhase` to fail when `find . -name additionalFile.c` succeeds.

messemar replied:
Ah, now I get what you mean. But indeed removed files might be an issue using this desired approach.

I have to think about a solution first.

YellowOnion commented:
Does this require two copies of src? this doesn't seem to create genuine incremental builds, but require one "unaltered" source, that acts as a seed for the 2nd derivation. This can be problematic when you want to work on big projects like linux source code.

roberth commented:
> Does this require two copies of src?

Yes.

> this doesn't seem to create genuine incremental builds

That is correct. "True" incremental builds are a cross-cutting concern that goes all the way down to the build tool used inside the sandbox. As such, Nix can not be solely responsible for managing the state, making such a solution fragile, unless the build tool can delegate the state management to _recursive nix_, which is, still, rather restrictive and an invasive requirement. 
Incrementality can't be added in a transparent manner, so instead, this is an embodiment of an incremental build within the derivation paradigm, that makes the prior state explicit and does not provide any guarantees beyond the fact that the result is an incremental build from a specific version to another specific version.

@messemar Could you explain this in the Nixpkgs manual?

> This can be problematic when you want to work on big projects like linux source code.

If you buy into Nix, you have to be ok with "wasting" some space for the benefits of reproducibility, atomicity, etc.
There's two kinds of space waste: the temporary and the permanent. The first just causes garbage collection to run more frequently, the latter should not be affected by this solution.
When you're rebuilding, say, linux you are already copying the `src` to the store for each new iteration (unless you only change the expression), so presumably the garbage produced isn't much of a problem.
I guess I'm skipping over some subtleties, but I think this explanation is already getting quite long.

messemar commented:
> @messemar Could you explain this in the Nixpkgs manual?

Yes I can, when I have some spare time again I will. As well as continuing the implementation, of course :D

kaldonir commented on pkgs/build-support/build-incremental.nix:
```diff
@@ -0,0 +1,47 @@
+    '' + (old.installPhase or "") + ''
+      mkdir -p $incrementalBuildArtifacts
+      cp -r ./* $incrementalBuildArtifacts/
+    '';
```
Replacing the `installPhase` like this will break derivations that have no explicit `installPhase`, but `preInstall` or `postInstall` hooks. See https://nixos.org/manual/nixpkgs/stable/#sec-stdenv-phases:

> When overriding a phase, for example installPhase, it is important to start with runHook preInstall and end it with runHook postInstall, otherwise preInstall and postInstall will not be run. Even if you don’t use them directly, it is good practice to do so anyways for downstream users who would want to add a postInstall by overriding your derivation.

Would it make sense to copy the build artifacts in the `pre-` or `postInstall` hook instead, to avoid this pitfall?

roberth replied:
In fact, this PR could be implemented by registering a `pre`- or `postInstall` hook instead of doing this.

@messemar Did you consider cutting off the entire build at the install phase, and having only the incremental build artifacts as an output? This would avoid problems caused by a non-idempotent original `installPhase`, such as `mv foo $out/`. It also wastes less space, because the original outputs aren't actually used anyway.

messemar replied:
But yes, as its for development use only we can have a cut off after the install phase too.

messemar committed:
checkpointedBuild: consider removing files and make buildartifacts the only output of the prepare step

messemar committed:
checkpointedBuild: fix tests for checkpointedBuild functions

messemar commented:
Steps to un-draft: Add a section in the nixpkgs manual.

messemar committed:
doc: add section about checkpointed build

dasJ commented on doc/builders/special/incremental-build.section.md:
```diff
@@ -0,0 +1,20 @@
+# pkgs.buildIncremental.*  {#sec-incremental-build}
+
+`pkgs.buildIncremental` provides a way to build derivations incrementally. It consists of two functions to make incremental builds using nix possible.
+**Attention** Incrementality can't be added in a transparent manner, so incrementality is added within the derivation paradigm. Nix is not able to manage the state, but the underlying build tool can delegate the state management to recursive nix.
```
I don't really understand what this is saying

roberth replied:
Perhaps this could be rephrased entirely in terms of state. Something like this?

```suggestion

For hermeticity, Nix derivations do not allow any state to carry over between builds, making a transparent incremental build impossible without turning each small build step into a derivation.

However, we can tell Nix explicitly what the previous build state was, by representing that previous state as a derivation output. This allows the passed build state to be used for an incremental build.
```

If we explain it, there's no need for extra caution.

roberth replied:
(edited)

asymmetric replied:
@roberth your explanation is much clearer.

dasJ commented on pkgs/build-support/build-incremental.nix:
```diff
@@ -0,0 +1,54 @@
+    * The build output can be used later to allow incremental builds
+    * by passing the derivation output to the `mkIncrementalBuild` function.
+    *
+    * To build a project incrementaly follow these steps:
```
Maybe this more concrete example should go into the manual?

messemar replied:
I have integrated it into the manual.

dasJ commented on pkgs/build-support/build-incremental.nix:
```diff
@@ -0,0 +1,54 @@
+      set -e
+      shopt -s extglob
+      rm -r !("sourceDifference.patch")
+      ls -al .
```
Is this still useful?

dasJ commented on pkgs/top-level/all-packages.nix:
```diff
@@ -309,6 +309,8 @@ with pkgs;
 
   buildMaven = callPackage ../build-support/build-maven.nix {};
 
+  buildIncremental = callPackage ../build-support/build-incremental.nix {};
```
I think something like `incrementalBuildTools` would be more suitable since this is not really a package by itself.

dasJ commented:
@roberth PTAL, I don't see any reasons apart from the ones I commented to not merge this. Is this fine by you?
Also cc @infinisil since you know a lot about things :)

infinisil commented on doc/builders/special/incremental-build.section.md:
```diff
@@ -0,0 +1,20 @@
+`pkgs.buildIncremental` provides a way to build derivations incrementally. It consists of two functions to make incremental builds using nix possible.
+**Attention** Incrementality can't be added in a transparent manner, so incrementality is added within the derivation paradigm. Nix is not able to manage the state, but the underlying build tool can delegate the state management to recursive nix.
+
+## Building a project incrementally 
```
This documentation needs to be clearer about how it can be used. It should come with an example showing users how it can be set up for a specific nixpkgs package, with code and commands that can be run and showing how much faster the build becomes.

messemar replied:
i've added the example from the code.

infinisil commented on pkgs/test/incrementalBuild/default.nix:
```diff
@@ -0,0 +1,57 @@
+  incrementalBuiltHelloWithRemovedFile = buildIncremental.mkIncrementalBuild patchedHelloRemoveFile baseHelloRemoveFileArtifacts;
+in
+stdenv.mkDerivation {
+  name = "patched-hello-returns-correct-output";
```
I think the tests should be more extensive. This is a very generic function that can be applied to essentially any derivation. Does it work with `chromium`? What about Python packages or Haskell packages? Does it work with all derivations perhaps? Does it work with derivations that mess around with phases? Etc.

Gabriella439 replied:
RE: Haskell packages, you can compare and contrast this PR with my analogous Haskell-specific PR in https://github.com/NixOS/nixpkgs/pull/204020

The one main thing that would not work if you applied this to Haskell builds is that macOS builds would not correctly work.  The reason why is because of the linker work-around on macOS (and the Haskell PR disables the linker workaround when incremental builds are enabled)

I think other than I think that the two approaches could be unified.  The other primary difference between the two is that the Haskell approach only exports the `dist/build` directory and not the entire directory, for efficiency reasons (since there is no benefit in reusing other build products)

messemar replied:
> The other primary difference between the two is that the Haskell approach only exports the dist/build directory and not the entire directory, for efficiency reasons (since there is no benefit in reusing other build products)

The export of the entire directory is essential in my use case, as some projects like the Linux Kernel  does not have a build directory, where all the build outputs are stored.

But yes, if there is a chance for unification, we should try it.

messemar replied:
> I think the tests should be more extensive. This is a very generic function that can be applied to essentially any derivation. Does it work with `chromium`? What about Python packages or Haskell packages? Does it work with all derivations perhaps? Does it work with derivations that mess around with phases? Etc.

I understand, why more extensive tests are fine.

In terms of build time, larger tests, such as a Linux kernel build will increase the testing time massively.

I can introduce such tests as well, but i would argue that large tests that tests the same behavior like the small ones are not that beneficial.

In terms of python, or haskell, I have not the knowledge about the build systems that i can argue it works for them as well.

infinisil reviewed (dismissed):
Oh interesting!

infinisil commented on pkgs/top-level/all-packages.nix:
```diff
@@ -309,6 +309,8 @@ with pkgs;
 
   buildMaven = callPackage ../build-support/build-maven.nix {};
 
+  buildIncremental = callPackage ../build-support/build-incremental.nix {};
```
I'm not sure about the term "incremental" being used. It's neat that this works, but it's not truly incremental as @roberth already explained. Using "incremental" sets the wrong expectations. Alternatives I can think of:
- Checkpointed builds
- Resumed builds
- Additive builds

messemar replied:
Ive renamed it to checkpointedBuildTools

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixpkgs-support-for-incremental-haskell-builds/24115/2

Gabriella439 commented:
I want to mention that I have a PR open against Nix which would make it possible to detect the earlier build automatically:

https://github.com/NixOS/nix/pull/7362

I also have a blog post explaining this in more detail:

https://www.haskellforall.com/2022/12/nixpkgs-support-for-incremental-haskell.html

messemar committed:
checkpointedBuilds: allow dotglob and remove debugging leftover

messemar committed:
checkpointedBuilds: add comments in the code

messemar commented on pkgs/build-support/build-incremental.nix:
```diff
@@ -0,0 +1,54 @@
+    installPhase = ''
+      mkdir -p $out/outputs
+      cp -r ./* $out/outputs/
+    '';
```
Fixed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixpkgs-support-for-incremental-haskell-builds/24115/7

messemar committed:
rename: incremental -> checkpointed builds

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixpkgs-support-for-incremental-haskell-builds/24115/8

messemar commented:
@roberth I would love to see it merged! I've just decided to make it a [standalone project](https://github.com/messemar/checkpoint-build-nix), to make it available to others in some way.

messemar committed:
checkpointBuild: fix whitespaces in documentation

parthy commented:
This would be brilliant to have available as a tool. Long-running builds from scratch due to tiny source changes are a major headache at times, and to be able to use this kind of compromise when it makes sense would be a huge improvement in daily work. I‘m looking forward to see this go in!

messemar committed:
doc: fix typo in checkpoint build section

Co-authored-by: Markus Partheymueller <markus.partheymueller@cyberus-technology.de>

dasJ commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -0,0 +1,67 @@
+    # We copy the complete build folder, to take care for
+    # Build tools, building in the source directory, instead of
+    # having a build root directory, e.G the Linux kernel.
+    installPhase = ''
```
You could also run the preInstall and postInstall hooks here? Or maybe other hooks in case you don't want to re-use these names. Something like `preCheckpointInstall` to make this extensible.

messemar replied:
fixed.

roberth reviewed (approved):
I don't want to block this on docs editing. Nonetheless, docs are an important area for the ecosystem to improve on, and hence the suggestions.

messemar committed:
doc: checkpointBuild: fix wording

Co-authored-by: Robert Hensing <roberth@users.noreply.github.com>

messemar committed:
checkpointBuilds: add hooks for checkpoint builds

messemar committed:
checkpointBuild: doc remove textual description in favor of an integrated example

messemar commented on doc/build-helpers/special/checkpoint-build.section.md:
```diff
@@ -0,0 +1,26 @@
+    * enjoy shorter build times
+
+As Nix has no builtin support for the detection of the previous built derivation, a base version needs to be declared.
+To create the outputs later used as base version for checkpoint builds, the function `pkgs.checkpointBuildTools.prepareCheckpointBuild` is used.
```
I've added an integrated example.

dasJ reviewed (approved):
Code looks good and functionality is lovely to have (would have helped me bisecting the kernel a couple of days ago :)

phip1611 commented on doc/build-helpers/special/checkpoint-build.section.md:
```diff
@@ -0,0 +1,36 @@
+# pkgs.checkpointBuildTools  {#sec-checkpoint-build}
+
+`pkgs.checkpointBuildTools` provides a way to build derivations incrementally. It consists of two functions to make checkpoint builds using nix possible.
```
nit:

```suggestion
`pkgs.checkpointBuildTools` provides a way to build derivations incrementally. It consists of two functions to make checkpoint builds using Nix possible.
```

phip1611 commented on doc/build-helpers/special/checkpoint-build.section.md:
```diff
@@ -0,0 +1,36 @@
+
+`pkgs.checkpointBuildTools` provides a way to build derivations incrementally. It consists of two functions to make checkpoint builds using nix possible.
+
+For hermeticity, Nix derivations do not allow any state to carry over between builds, making a transparent incremental build within a derivation impossible.
```
not sure about this one, but IMHO it sounds more natural

```suggestion
For hermeticity, Nix derivations do not allow any state to be carried over between builds, making a transparent incremental build within a derivation impossible.
```

messemar committed:
doc: checkpointBuild: fix wording

Co-authored-by: Philipp Schuster <phip1611@gmail.com>

messemar commented:
@roberth @dasJ  Is there anything left, that I can do to get this over the finishing line?

dasJ commented:
There isn't

dasJ commented:
Ah there is: @infinisil is it okay to dismiss your review?

phip1611 commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -0,0 +1,69 @@
+    outputs = [ "out" ];
+    name = drv.name + "-checkpointArtifacts";
+    # To determine differences between the state of the build directory
+    # from an earlier build and  a later one we store the state of the build
```
nit:

```suggestion
    # from an earlier build and a later one we store the state of the build
```

phip1611 commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -0,0 +1,69 @@
+      cp -r ./* $out/sources/
+    '';
+
+    # After the build the build directory is copied again
```
```suggestion
    # After the build, the build directory is copied again
```

phip1611 commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -0,0 +1,69 @@
+    # After the build the build directory is copied again
+    # to get the output files.
+    # We copy the complete build folder, to take care for
+    # Build tools, building in the source directory, instead of
```
```suggestion
    # build tools, building in the source directory, instead of
```

phip1611 commented on pkgs/build-support/checkpoint-build.nix:
```diff
@@ -0,0 +1,69 @@
+    # to get the output files.
+    # We copy the complete build folder, to take care for
+    # Build tools, building in the source directory, instead of
+    # having a build root directory, e.G the Linux kernel.
```
```suggestion
    # having a build root directory, e.g., the Linux kernel.
```

phip1611 reviewed (approved):
Awesome! Left some nits.

infinisil commented:
> Ah there is: @infinisil is it okay to dismiss your review?

Oh yeah sorry missed this. It's okay. Generally this does seem a rather hacky, but it looks like it's good enough for a lot of use cases, so it's fine.

@messemar: Though it looks like https://github.com/NixOS/nixpkgs/pull/167670#discussion_r1052295282 wasn't fully addressed, and I can't find a comment explaining why.

roberth commented:
> rather hacky

I'm sure you think the same about it, but to be clear, the goal itself is hacky, as something non-hacky would have to integrate with something other than phases perhaps, in order for it to be able to give guarantees that can't be messed up by the user of this feature, but it's unclear to what that would be.
The implementation is not at all hacky in and of itself.

messemar commented:
@infinisil and @roberth  
I totally agree on the hackyness. But as checkpoints and incremental builds are not a first class concept for nix, it can not be enabled in a non hacky way, I think.


Regarding https://github.com/NixOS/nixpkgs/pull/167670#discussion_r1052295282 If we use `dotglob` wouldn't that mean that we also remove ´.´ and ´..´ as well ? Does not sound like a good Idea IMHO. But yes, we might miss dotfiles then.

infinisil commented:
@messemar You do have `dotglob` enabled in the PR. I was thinking of the other point in that comment that allows avoiding the `extglob` for `rm -r !("sourceDifference.patch")`.

messemar commented:
@infinisil Ah,  now I get it.  Is it a nit, or should I fix it in a follow up PR ?

infinisil commented:
@messemar I'd classify it as a nice-to-have for correctness. Feel free to make a follow-up PR, but not necessarily :)

asymmetric commented:
Would it make sense to mention this in the 2023 recap?

roberth commented:
@asymmetric It needs an explicit starting state, and it only works for typical `mkDerivation` based builds. That should be made very clear. I would also like to see it used in practice somewhere.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/idea-mkderivation-auto-generate-unwrapped-package-right-after-build/37901/3

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/idea-mkderivation-auto-generate-unwrapped-package-right-after-build/37901/1

bryango commented:
I really, really like this idea and would like to use it in practice! However, there are some minor inconsistencies across the docs; e.g. `mkCheckpointBuild` is mentioned in the doc:

https://github.com/NixOS/nixpkgs/blob/d6b045fcc3d2eb036e16b04b1127f1457eb242d7/doc/build-helpers/special/checkpoint-build.section.md#L34-L36

But the actual command is `mkCheckpointedBuild` with an `ed`.

In fact, all other functions in this PR are in the form of `*CheckpointBuild*` without the `ed`, so I think we might as well trade the `mkCheckpointedBuild` name in favor of `mkCheckpointBuild` for an uniformed look (and link `mkCheckpointedBuild` to `mkCheckpointBuild` with a `lib.warn "deprecated"`, although [currently there is no usage of this in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+mkCheckpointedBuild&type=code)). Is that okay? If so, I can make a PR!

messemar commented:
> In fact, all other functions in this PR are in the form of *CheckpointBuild* without the ed, so I think we might as well trade the mkCheckpointedBuild name in favor of mkCheckpointBuild for a uniformed look (and link mkCheckpointedBuild to mkCheckpointBuild with a lib.warn "deprecated", although [currently there is no usage of this in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+mkCheckpointedBuild&type=code)). Is that okay? If so, I can make a PR!

@bryango From my perspective, the change is reasonable and I like it.
